memtable/skiplistrep.cc

```diff
@@ -95,6 +95,37 @@ class SkipListRep : public MemTableRep {
     return (end_count >= start_count) ? (end_count - start_count) : 0;
   }
 
+  void UniqueRandomSample(const uint64_t& num_entries,
+                          const uint64_t& sample_size,
+                          std::unordered_set<const char*>* entries) override {
+    entries->clear();
+    SkipListRep::Iterator iter(&skip_list_);
+    // If the sample size is at least 25% of the total number of elements
+    // in the skip list, then simply iterate linearly over the list.
+    // (25% is arbitrary, could sample_size>N/log(N) if we assume iterating
+    // over each sample is O(N) and RandomSeek calls would be
+    // O(sample_size*log(N))).
+    if (sample_size >= (num_entries / 4)) {
+      Random rnd(1234);
+      int n = (sample_size > num_entries)
+                  ? 1
+                  : static_cast<int>(num_entries / sample_size);
+      for (iter.SeekToFirst(); iter.Valid(); iter.Next()) {
+        if (rnd.OneIn(n)) {
+          entries->insert(iter.key());
+        }
+      }
+    } else {
+      for (uint64_t i = 0; i < sample_size; i++) {
+        do {
+          iter.RandomSeek();
+        } while (!iter.Valid() ||
+                 (entries->find(iter.key()) != entries->end()));
+        entries->insert(iter.key());
```

Comment:
I assumed the unordered_set was so that we didn't put the same (by pointer) key in more than once if we happen to sample it more than once. But the loop bound is currently number of items inserted, not number of items ending up in `entries`. (I might be suggesting violating my own rule about risking infinite loops. I'll see what you come up with.)

Reply:
Here is my suggestion. I actually plugged some numbers using the dynamic thresholding, and when we choose this sampling method the likelihood of inserting twice the same element are completely negligible. (I added more comments in there). 
In brief, the probability of having at least one duplicate is P = 1-(N!/((N-m)! * N^m)) (N # of elements in table, and m the `target_sample_size`). For N=1000, since the max `m` is going to be m=1 ( because sqrt(N)/10 = 1), we get P=0.1% . This value goes up to 0.5% for N=10**10 entries, which wont ever happen.
So I am not checking anything anymore, and the unsorted_set will take care of potential duplicates. I am also renaming the variable 'target_sample_size' instead of 'sample_size' because it is not strictly enforced (note that it isn't strictly enforced in the other sampling method either). I am also making this clear in the comments.

Reply:
I am now giving it 5 attempts to find a non-duplicate entry. 
At worst, for the final pick , when m=N/10 there is 10% chances to find a duplicate. So chances to fail finding a non-duplicate after 5 attempts is 0.001%.